src/grid/storage.rs

```diff
@@ -34,86 +36,81 @@ impl<T: PartialEq> ::std::cmp::PartialEq for Storage<T> {
 
 impl<T> Storage<T> {
     #[inline]
-    pub fn with_capacity(cap: usize, lines: Line) -> Storage<T> {
+    pub fn with_capacity(cap: usize, lines: Line, template: T) -> Storage<T>
+    where
+        T: Clone,
+    {
+        // Allocate all lines in the buffer, including scrollback history
+        //
+        // TODO (jwilm) Allocating each line at this point is expensive and
+        // delays startup. A nice solution might be having `Row` delay
+        // allocation until it's actually used.
+        let inner = vec![template; cap];
         Storage {
-            inner: Vec::with_capacity(cap),
+            inner,
             zero: 0,
             visible_lines: lines - 1,
+            len: cap,
         }
     }
 
-    #[inline]
-    pub fn capacity(&self) -> usize {
-        self.inner.capacity()
-    }
-
     /// Increase the number of lines in the buffer
     pub fn grow_visible_lines(&mut self, next: Line, template_row: T)
     where
         T: Clone,
     {
-        // Calculate insert position (before the first line)
-        let offset = self.zero % self.inner.len();
-
-        // Insert new template row for every line grown
+        // Number of lines the buffer needs to grow
         let lines_to_grow = (next - (self.visible_lines + 1)).0;
-        for _ in 0..lines_to_grow {
-            self.inner.insert(offset, template_row.clone());
-        }
 
-        // Set zero to old zero + lines grown
-        self.zero = offset + lines_to_grow;
+        // Only grow if there are not enough lines still hidden
+        if lines_to_grow > (self.inner.len() - self.len) {
+            // Lines to grow additionally to invisible lines
+            let new_lines_to_grow = lines_to_grow - (self.inner.len() - self.len);
 
-        // Update visible lines
-        self.visible_lines = next - 1;
-    }
-
-    /// Decrease the number of lines in the buffer
-    pub fn shrink_visible_lines(&mut self, next: Line) {
-        // Calculate shrinkage and last line of buffer
-        let shrinkage = (self.visible_lines + 1 - next).0;
-        let offset = (self.zero + self.inner.len() - 1) % self.inner.len();
+            // Get the position of the start of the buffer
+            let offset = self.zero % self.inner.len();
 
-        // Generate range of lines that have to be deleted before the zero line
-        let start = offset.saturating_sub(shrinkage - 1);
-        let shrink_before = start..(offset + 1);
+            // Split off the beginning of the raw inner buffer
+            let mut start_buffer = self.inner.split_off(offset);
 
-        // Generate range of lines that have to be deleted after the zero line
-        let shrink_after = (self.inner.len() + offset + 1 - shrinkage)..self.inner.len();
+            // Insert new template rows at the end of the raw inner buffer
+            let mut new_lines = vec![template_row; new_lines_to_grow];
+            self.inner.append(&mut new_lines);
```

Comment:
A future optimization here will be to grow the underlying storage, shift existing elements, and insert new elements -- all without the vec split.

Reply:
Wasn't able to figure out how to make this work, but it seems like a good idea.